### ISSUE TITLE
Hidden recipient message decryption problem

### ISSUE THREAD

**Issue by kazetkazet** [openpgpjs/openpgpjs#392]
I generate a key via openpgp.generateKeyPair({numBits: 2048, userId: 'a@xxxx.net', passphrase: ''}).

Then, I take its publicKeyArmored and import in gnupg and run the following:

echo "blah blah" | gpg -u MY_KEY -se --armor --hidden-recipient NEWLY_GENERATED_KEY

Then I get "TypeError: Cannot read property 'isDecrypted' of null" when trying to run:

openpgp.decryptMessage(NEWLY_GENERATED_KEY.key, openpgp.message.readArmored(MESSAGE)) where MESSAGE is the message created by gnupg.

What am I doing wrong? When I replace --hidden-recipient with -r in gnupg invocation everything works fine.

**Comment by wgromniak** [comment]
I'm not a JavaScript programmer, but looking at [message.js](https://github.com/openpgpjs/openpgpjs/blob/master/src/message.js#L92), function `decrypt`, lines 92-108:
```javascript
var encryptionKeyIds = this.getEncryptionKeyIds();
  if (!encryptionKeyIds.length) {
    // nothing to decrypt return unmodified message
    return this;
  }
  var privateKeyPacket = privateKey.getKeyPacket(encryptionKeyIds);
  if (!privateKeyPacket.isDecrypted) throw new Error('Private key is not decrypted.');
  var pkESKeyPacketlist = this.packets.filterByTag(enums.packet.publicKeyEncryptedSessionKey);
  var pkESKeyPacket;
  for (var i = 0; i < pkESKeyPacketlist.length; i++) {
    if (pkESKeyPacketlist[i].publicKeyId.equals(privateKeyPacket.getKeyId())) {
      pkESKeyPacket = pkESKeyPacketlist[i];
      pkESKeyPacket.decrypt(privateKeyPacket);
      break;
    }
  }
```
we look at all the key ids of the encrypted packets, then take the first subkey of given `privateKey` matching one of the ids and decrypt the packet with the key id matching the id of the subkey, if I understand this correctly.

Whereas, in [RFC 4880, Section 5.1](http://tools.ietf.org/html/rfc4880#section-5.1) we read:

>   An implementation MAY accept or use a Key ID of zero as a "wild card"
>  or "speculative" Key ID.  In this case, the receiving implementation
>   would try all available private keys, checking for a valid decrypted
>   session key.  This format helps reduce traffic analysis of messages.

The code above will fail to decrypt a packet with 0 key id (a subkey with 0 key id does not exist). To make it work, we could try decryption with all the keys, until a working one (i.e. the decryption does not throw errors, completes successfully, &.) is found, if a packet with 0 key id is encountered. We already pass the private key to the decrypt function, so trying all its subkeys is almost costless.

As I understand, this feature is optional, but it would be nice, if openpgpjs supported it, as it nicely enhances privacy.

**Comment by sanjanarajan** [comment]
Fixed by #635
